client/state/data-layer/http-data.js

```diff
@@ -184,8 +199,70 @@ export const requestHttpData = ( requestId, fetchAction, { fromApi, freshness =
 	return data;
 };
 
+/**
+ * Blocks execution until requested data has been fulfilled
+ *
+ *  - May return without data if data hasn't been fulfilled or failed
+ *  - Use for SSR contexts or when we _want_ to block rendering or execution
+ *    until the requested data has been fulfilled, e.g. when URL routing
+ *    depends on some data property
+ *  - _DO NOT USE_ when normal synchronous/data interactions suffice such
+ *    as is the case in 99.999% of React component contexts
+ *
+ * @example:
+ * withData( {
+ *     geo: () => requestGeoLocation(),
+ *     splines: () => requestSplines( siteId ),
+ * } ).then( ( { geo, splines } ) => {
+ *     return ( geo.state === 'success' || splines.state === 'success' )
+ *         ? res.send( renderToStaticMarkup( <LocalSplines geo={ geo.data } splines={ splines.data } /> ) )
+ *         : res.send( renderToStaticMarkup( <UnvailableData /> ) );
+ * }
+ *
+ * @param {object} query key/value pairs of data name and request
+ * @param {int} timeout how many ms to wait until giving up on requests
+ * @return {Promise<object>} fulfilled data of request (or partial if could not fulfill)
+ */
+export const withData = ( query, { timeout = 5000 } = {} ) =>
+	new Promise( ( resolve, reject ) => {
+		let unsubscribe = () => {};
+		let timer = null;
+		const names = Object.keys( query );
+
+		const getValues = () =>
+			names.reduce(
+				( [ values, allGood ], name ) => {
+					const value = query[ name ]();
```

Comment:
> but what you really want is to call a pure selector

maybe? I'd say it's something to argue about which tradeoffs are more important. if we have a slow data request and a fast data request with short freshness then if we only query here then by the time the slow request resolves the fast request will be stale.

> At this moment it works, because the query is pending after issued for the first time and won't be issued on the subsequent calls etc. But one day, as features and options are added, it won't work any more.

maybe in another design but this is intentionally an imperative mutable API and if we re-issue a call while one is `pending` then it means we've introduced an error into the system and should not merge that patch (or fix the bug).

Reply:
> if we have a slow data request and a fast data request with short freshness then if we only query here then by the time the slow request resolves the fast request will be stale.

Then the process of "waiting for 2+ requests and requiring that all results are fresh" doesn't always have a well-defined result. Response 1 can become stale while still waiting for response 2, so we reissue request 1. While waiting for response 1-try2, response 2 becomes stale and we reissue request 2. Leads to starvation.

"very short freshness" looks like a very hypothetical and unreasonable edge case, until we introduce a [`network-only` fetch policy](https://www.apollographql.com/docs/react/api/react-apollo.html#graphql-config-options-fetchPolicy). That means we never want cached data and always issue a network request. Equivalent to freshness being zero, which is a very short value indeed.

We are not going to solve this today in this PR, just writing it down for later.